File: Src/Runtime/Dictionaries/Access.cpp
/*
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #                           Typed                     # *
 * #                   A text formatting DSL             # *
 * #                                                     # *
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #         Created by Rodrigo R. & Contributors        # *
 * #         Released under the Apache License 2.0       # *
 * #            Check LICENSE.MD for more info           # *
 * #                                                     # *
 * #-----------------------------------------------------# *
*/

//
// Created by Rodrigo on 5/25/26.
//

#include "Access.h"

#include "ADT/Exception/UnknownElement.h"
#include "ADT/Lang/AST.h"
#include "ADT/Runtime/Object.h"

using namespace Typed;
using namespace Typed::Runtime;
using namespace Typed::Runtime::Dictionaries;

ADT::Runtime::Object Dictionaries::Access(
    ADT::List::Object &args,
    ADT::Lang::AST *trace
)
{
    auto &dict = Support::Runtime::GetDictionaryObj(args[0]);
    auto &key = args[1];

    auto it = dict->find(key);
    if (it == dict->end())
    {
        throw ADT::Exception::UnknownElement(
            trace->line,
            trace->column
        );
    }

    return it->second;
}

File: src/runtime/dictionaries/access.rs
/*
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #                           Typed                     # *
 * #                   A text formatting DSL             # *
 * #                                                     # *
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #         Created by Rodrigo R. & Contributors        # *
 * #         Released under the Apache License 2.0       # *
 * #            Check LICENSE.MD for more info           # *
 * #                                                     # *
 * #-----------------------------------------------------# *
*/
use crate::adt::error::RuntimeError;
use crate::adt::lang::{RuntimeArguments, AST};
use crate::adt::result::ExecutionTupleResult;
use crate::adt::runtime::Object;
use crate::support::runtime::execution::continue_execution;
use crate::support::runtime::object::get_dictionary;

pub fn access<'a>(
    args: RuntimeArguments<'a>,
    trace: &AST<'a>
)
    -> ExecutionTupleResult<'a>
{
    let dict_obj = args.get(0).unwrap();
    let key = args.get(1).unwrap();
    
    if let Object::NonHashable(_) = key {
        return Err(
            RuntimeError::key_not_found(trace)
        );
    }

    let 
        Object::Hashable(key) 
    = key else { unreachable!() };
    let dict = get_dictionary(dict_obj, trace)?;

    let dict = dict.borrow();
    let value = dict.get(&key);
    
    if value.is_none() {
        return Err(
            RuntimeError::key_not_found(trace)
        );
    }
    
    continue_execution(value.unwrap().clone())
}